File: 16-3sum-closest/16-3sum-closest.cs
public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        int l = nums.Length;
        int r = nums[0]+nums[1]+nums[l-1];
        Array.Sort(nums);
        for(int i=0;i<l-1;i++){
            if(i==0 || (i>0 && nums[i]!=nums[i-1])){
                int lo=i+1,hi=l-1;
                while(lo<hi){
                    int x= nums[i]+nums[lo]+nums[hi];
                    if(x>target){
                        hi--;
                    }else{
                        lo++;
                    }
                    if(Math.Abs(x-target)<Math.Abs(r-target)){
                        r=x;
                    }
                }
            }
        }
        return r;
    }
}

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        
        int n = nums.size();
        int res = nums[0] + nums[1] + nums[2];
        
        for (int i = 0; i < n-2; i++)
        {
            int l = i+1, r = n-1;
            while (l<r)
            {
                int curr = nums[l]+nums[r]+nums[i];
                if(abs(curr-target)<abs(res-target))
                {
                    res=curr;
                }
                else if(curr==target) break;
                
                if(curr>target){
                    r--;
                }else{
                    l++;
                }
            }                        
        }        
        return res;
    }
};